### DISCUSSION TITLE
Weighting unit in 2D

### DISCUSSION THREAD

**Discussion by pordyna** [BLAST-WarpX/warpx#4712] [Q&A] [answered]
It looks to me that the particle weight in an output from a 2D simulation (as well as the accumulated value in a particle histogram) is actually not the particle number, but rather linear particle density along the 3rd not simulated direction. To be honest, this makes a lot of sense, but in the `openPMD` output the weighting is marked as dimensionless. Could this be a bug or is it intentional?

**Comment by pordyna** [comment]
What also could happen is that I was supposed to set somewhere the cell size in the not simulated dimension but didn't, and it was set to 1m by default? 

Btw. I'm actually not sure of the `openPMD` `unitDiemension` for the weighting attribute (can check it once perlmutter is back online), but I had a particle field defined as:
`fields.particle_fields.particle_number(x,y,z,ux,uy,uz) = 1`  and it has a unitless unitDimension in the output. While it seems to be either a linear density (or a number of particles in a 1m long cell :D in that case dimensionless is correct)
Here is more context for the field output: 
```
fields.diag_type = Full
fields.fields_to_plot = By Ex Ez
fields.file_min_digits = 8
fields.format = openpmd
fields.intervals = ::26631,57587:73566:533,36282:57587:5326,62913:62913:,73566::5326,5656:17640:533,9650:9650:,17640:36282:5326
fields.openpmd_backend = bp
fields.particle_fields.average_gamma(x,y,z,ux,uy,uz) = sqrt(1 + ux**2+uy**2+uz**2)
fields.particle_fields.average_gamma.do_average = 1
fields.particle_fields.average_momentum_x(x,y,z,ux,uy,uz) = ux
fields.particle_fields.average_momentum_x.do_average = 1
fields.particle_fields.average_momentum_z(x,y,z,ux,uy,uz) = uz
fields.particle_fields.average_momentum_z.do_average = 1
fields.particle_fields.particle_number.do_average = 0
fields.particle_fields_to_plot = average_gamma average_momentum_x average_momentum_z particle_number
fields.write_species = 0
```

**Reply by ax3l** [reply, marked as answer]
Hi @pordyna,

Thanks for the report. Yes, you got this right. Weighting is actually a number of particles (count, unitless) only in 3D, and in 2D a count per length and in 1D a count per surface area.

We need to change this in both WarpX:
- https://github.com/ECP-WarpX/WarpX/pull/4717
and openPMD-standard:
- https://github.com/openPMD/openPMD-standard/pull/285

Thanks for the report!!
